Zagrios/bs-manager
# [feat-560] ask user for installation folder/path

## Scope

[closes TICKET-560](https://github.com/Zagrios/bs-manager/issues/560)
[closes TICKET-481](https://github.com/Zagrios/bs-manager/issues/481)

## Implementation

Added a way to ask the user for an existing or non-existing installation folder/path

Modified `modal.component.tsx` to add the property **closable**. If **closable = false**, then you can't exit out the modal by any means, except by adding a button to resolve the modal itself.

Created `AskInstallPathModal` (ask-install-path-modal.component.tsx).

Created `SetupService` (setup.service.ts) class to handle any user configuration before application startup. Logic such as sequential modals/prompts can be added here sequentially to ask for config values. Not sure if one whole modal or separate modals will be done but should be easy enough to edit the logic there.

Created `bs-installer-ipcs.ts` to handle all the server calls for the installation folder/path. Transferred some codes from `bs-download-ipcs.ts` to here.

Might need to transfer the the installation folder/path code from `SteamDownloaderService` to another class as well, but might do it in another PR.

## Screenshots

![image](https://github.com/user-attachments/assets/e1fe89ab-d6d9-4ba3-b92a-f4b39801fd68)

*UI may still change depending on suggestions*
As of commit 9e2f716c53d61eccc8d9e247de4af8d415887567

## How to Test

### Test Case

1. Edit the **installation-folder** property in the `config.json` file with any existing or non-existing folder path other than your original installation folder.
- Windows: `%AppData%/Roaming/bs-manager`
- Linux: `/.config/bs-manager`
2. Open BSManager and a modal will appear (similar to the above screenshot)
3. Select any folder and click on **Confirm**
4. On close either:
4.1. If existing installation folder: version should be shown on the sidepanel
4.2. If non-existing folder: installation folder will be created

### Additional Tests
- Check if the **Default** button works properly.
- Check if this would not ask the user if the installation path points to an existing installation folder.
- Check if the songDetailsCache was loaded properly.

## New/Updated Resource Text

- [x] **misc.confirm:** Confirm
- [x] **pages.settings.installation-folder.description:** Change the folder that will contain all the content downloaded by BSManager.
- [x] **modals.ask-install-path.title:** Installation folder
- [x] **modals.ask-install-path.choose-folder-description:** Choose the folder that will contain all the content downloaded by BSManager.
- [x] **modals.ask-install-path.choose-folder:** Choose folder
- [x] **modals.ask-install-path.default:** Default
- [x] **modals.ask-install-path.default-tooltip:** Defaults to your home folder

**NOTE:** ticked are already translated to other languages.

## Emoji Guide

**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**

E.g: Praise, minor suggestions, or clarifying questions that don’t block merging the PR.

> 🟢 Nice refactor!

> 🟡 Why was the default value removed?

E.g: Blocking feedback must be addressed before merging.

> 🔴 This change will break something important

|              |                |                                     |
| ------------ | -------------- | ----------------------------------- |
| Blocking     | 🔴 ❌ 🚨       | RED                                 |
| Non-blocking | 🟡 💡 🤔 💭    | Yellow, thinking, etc               |
| Praise       | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |
